std/range/package.d

```diff
@@ -8578,6 +8578,529 @@ if (isForwardRange!Source && hasLength!Source)
+are shortened to match the length of the source range.
+
+If the `source` range is longer than the last index provided, all extra elements will
+form one new chunk.
+
+If `Source` and `Partitioners` are forward ranges, the resulting range will be a forward range as well.
+Otherwise, the resulting chunks will be input ranges using the same input. The chunk will shrink
+with every `front` call and `popFront` will invalidate the previous `front`.
+
+Params:
+    source = range from which the chunks will be selected
+    partitioners = range specifying how the source range will be split into chunks
+
+Returns: Range of chunks of variable lengths
+*/
+auto chunks(Partitioning partitionType = Partitioning.ByLength, Source, Partitioners)
+    (Source source, Partitioners partitioners)
+if (isInputRange!Source && isInputRange!Partitioners && is(ElementType!Partitioners : size_t))
+{
+    static struct Chunks
+    {
+        static if (isForwardRange!Source && isForwardRange!Partitioners)
+        {
+            this(Source source, Partitioners partitioners, size_t offset = 0, UseMaxLength maxLen=UseMaxLength(false))
+            {
+                import std.algorithm.sorting : isStrictlyMonotonic;
+                static if (partitionType == Partitioning.ByIndex)
+                    assert(partitioners.isStrictlyMonotonic, "The indexes must be in a strictly increasing order");
+
+                _source = source;
+                _partitioners = partitioners;
+                _offset = offset;
+                _maxLen = maxLen;
+                if (!_source.empty)
+                    _isEmpty = false;
+                else
+                    _isEmpty = true;
+            }
+
+            /// Input range primitives. Always present.
+            @property bool empty()
+            {
+                return _isEmpty;
+            }
+
+            /// Ditto
+            @property auto front()
+            {
+                assert(!empty, "Attempting to fetch the front of an empty range");
+
+                if (_partitioners.empty && !_source.empty)
+                    return _source.save;
+
+                assert(_partitioners.front > _offset, "Attempting to create a chunk of length 0");
+
+                return _source.save.take(_partitioners.front - _offset);
+            }
+
+            /// Ditto
+            void popFront()
+            {
+                assert(!empty, "Attempting to popFront an empty range");
+
+                // source range is longer than index
+                if (_partitioners.empty && !_source.empty)
+                {
+                    _isEmpty = true;
+                    return;
+                }
+
+                assert(_partitioners.front > _offset);
+                _source.popFrontN(_partitioners.front - _offset);
+                static if (partitionType == Partitioning.ByIndex)
+                    _offset = _partitioners.front;
+                _partitioners.popFront();
```

Comment:
It's completely unclear to me how this takes into account the ByLength and ByIndex difference. Can you explain how this works? Does `_offset` stay 0 for the entire time if ByLength is used? If so, this is terrible design for saving a few static ifs (plus the compiler isn't going to be able to optimize out the subtraction).

I also don't understand why we seem to be forbidding zero-sized chunks.